pkg/cmd/rollbackcopy/rollbackcopy.go

```diff
@@ -0,0 +1,108 @@
+package rollbackcopy
+
+import (
+	"context"
+	"fmt"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"io"
+	"k8s.io/klog"
+	"math/rand"
+	"os"
+	"time"
+)
+
+type rollbackCopyOpts struct {
+	checkLeadershipInterval time.Duration
+	rollbackCopyInterval    time.Duration
+	configDir               string
+	errOut                  io.Writer
+}
+
+const (
+	leaderShipCheckInterval time.Duration = 5 * time.Minute
+	rollbackSaveInterval    time.Duration = 60 * time.Minute
+)
+
+func NewRollbackCopy(errOut io.Writer) *cobra.Command {
+	rollbackCopyOpts := &rollbackCopyOpts{
+		checkLeadershipInterval: leaderShipCheckInterval,
+		rollbackCopyInterval:    rollbackSaveInterval,
+		errOut:                  errOut,
+	}
+	cmd := &cobra.Command{
+		Use:   "rollbackcopy",
+		Short: "Periodically save snapshot and resources every hour, useful for a rollback",
+		Run: func(cmd *cobra.Command, args []string) {
+			must := func(fn func() error) {
+				if err := fn(); err != nil {
+					if cmd.HasParent() {
+						klog.Fatal(err)
+					}
+					fmt.Fprint(rollbackCopyOpts.errOut, err.Error())
+				}
+			}
+
+			must(rollbackCopyOpts.Run)
+		},
+	}
+	rollbackCopyOpts.AddFlags(cmd.Flags())
+	return cmd
+}
+
+func (r *rollbackCopyOpts) AddFlags(fs *pflag.FlagSet) {
+	fs.Set("logtostderr", "true")
+	fs.StringVar(&r.configDir, "config-dir", "/etc/kubernetes", "Dir containing kubernetes resources")
+}
+
+func checkAndScheduleBackup(memberName string, delay time.Duration, configDir string, done <-chan struct{}) {
+	backupTicker := time.NewTicker(delay)
+	for {
+		cli, err := getEtcdClient([]string{"localhost:2379"})
+		if err != nil {
+			klog.Error("Failed to get etcd client")
+			return
+		}
+		klog.Info(" checking amLeader in schedule...")
+		if !checkLeadership(cli, memberName) {
+			klog.Info("Member is NOT the leader. Returning")
+			cli.Close()
+			return
+		}
+		klog.Info("Member IS the leader. Backing up")
+		backup(cli, configDir)
+		cli.Close()
+
+		select {
+		case <-done:
+			backupTicker.Stop()
+			return
+		case <-backupTicker.C:
+			continue
+		}
+	}
+}
+
+func (r *rollbackCopyOpts) Run() error {
+	rand.Seed(time.Now().UnixNano())
+	ctx, cancel := context.WithCancel(context.TODO())
+	defer cancel()
+	localEtcdName := os.Getenv("ETCD_NAME")
+	ticker := time.NewTicker(r.checkLeadershipInterval)
+	go func() {
+		for {
+			checkAndScheduleBackup(localEtcdName, r.rollbackCopyInterval, r.configDir, ctx.Done())
+			select {
+			case <-ctx.Done():
+				ticker.Stop()
+				return
+			case <-ticker.C:
+				continue
+			}
+		}
+	}()
+
+	<-ctx.Done()
+	klog.Info("Done !!")
+	return nil
```

Comment:
This is a little difficult for me to follow... one way to unwind it might be like:

```go
// perform an initial backup
// TODO: only if necessary? or backup() should be idempotent
// with regards to time and period
backup()

ticker := time.NewTicker(period)
backupLoop:
for {
  select {
   case <-ticker.C:
     backup() // should no-op unless leader
   case <-signaled:
     break backupLoop
   case <-done:
     break backupLoop
  }
}
```

Also, we need to handle graceful shutdown on signal

Reply:
It was originally like that. But we wanted to it to run the very first time too. That's why I changed it to in that form based on this discussion:

https://stackoverflow.com/questions/32705582/how-to-get-time-tick-to-tick-immediately

Reply:
I can appreciate wanting uniformity of the call to backup, but I'm not really sure the complexity is worth it vs. just running it once before looping.

In any case, you need to handle shutdown signals

Reply:
Sure. I rewrote it.

Reply:
Still not quite there... I'm proposing the entirety of Run() be something like this:

```go
func Run(memberName, configDir string) error {
  // this should also probably be injected as a parameter
  done := make(chan os.Signal, 1)
  signal.Notify(c, os.Interrupt, syscall.SIGTERM)

  initial := make(chan struct{}, 1)
  initial <- struct{}{}

  ticker := time.NewTicker(period)

  backupLoop:
  for {
    select {
    case <-initial:
      backup(memberName, configDir) // should no-op unless leader
    case <-ticker.C:
      backup(memberName, configDir) // should no-op unless leader
    case <-done:
      break backupLoop
    }
  }

  return nil
}
```

Reply:
@retroflexer I don't see updates per @ironcladlou comments. re signals.

Reply:
OK. I will add signal.

Reply:
This loop is wrong. There are two different timer values, and hence we need two different functions to deal with the timers. I don't want to re-do the logic at this hour. I will just add the signal.

Reply:
Pushed a new commit.

Reply:
@ironcladlou pushed a new commit. Please take a look. Thanks.

Reply:
Looks like it should work now, thanks!